Sources/StreamChat/WebSocketClient/Engine/URLSessionWebSocketEngine.swift

```diff
@@ -65,46 +79,75 @@ class URLSessionWebSocketEngine: NSObject, WebSocketEngine, URLSessionDataDelega
             }
         }
     }
+
+    private func makeURLSessionDelegateHandler() -> URLSessionDelegateHandler {
+        let urlSessionDelegateHandler = URLSessionDelegateHandler()
+        urlSessionDelegateHandler.didOpenWith = { [weak self] _ in
+            self?.callbackQueue.async {
+                self?.delegate?.webSocketDidConnect()
+            }
+        }
+
+        urlSessionDelegateHandler.didCloseWith = { [weak self] closeCode, reason in
+            var error: WebSocketEngineError?
+
+            if let reasonData = reason, let reasonString = String(data: reasonData, encoding: .utf8) {
+                error = WebSocketEngineError(
+                    reason: reasonString,
+                    code: closeCode.rawValue,
+                    engineError: nil
+                )
+            }
+
+            self?.callbackQueue.async { [weak self] in
+                self?.delegate?.webSocketDidDisconnect(error: error)
+            }
+        }
+
+        urlSessionDelegateHandler.didCompleteWith = { [weak self] error in
+            // If we received this callback because we closed the WS connection
+            // intentionally, `error` param will be `nil`.
+            // Delegate is already informed with `didCloseWith` callback,
+            // so we don't need to call delegate again.
+            guard let error = error else { return }
+
+            self?.callbackQueue.async { [weak self] in
+                self?.delegate?.webSocketDidDisconnect(error: WebSocketEngineError(error: error))
+            }
+        }
+
+        return urlSessionDelegateHandler
+    }
+
+    deinit {
```

Comment:
Shouldn't the deinit be close (or above) the init?

Reply:
I usually use it at the end of the classes. In the current project, this is what we do more often, although in some places I've seen it closer to the `init`. For me I'm fine with both approaches if we keep it consistent across the SDK, maybe let's bring this discussion to a tech meeting to discuss this style guide? At least from what I can found this is not mentioned in any style guide.

Reply:
+1 on the Style Guide. We might need to start documenting some standards 👌 ! Let's ignore my comment for now, then